### PR TITLE
Faster balance sheets

### PR DESCRIPTION
**Pull request by xentac**
Two improvements, when mixed with https://bitbucket.org/blais/beancount/issues/191/beancountcoreinventory-should-be-a-dict, reduce my balance sheet computation time from over 3 seconds to around 1 second.

### CODE REVIEW

File: fava/core/charts.py [hunk 1]
```diff
@@ -165,11 +165,16 @@ def net_worth(self, interval):
         txn = next(transactions, None)
         inventory = Inventory()
 
+        today = datetime.datetime.today().replace(
```
Comment (yagebu) [inline fava/core/charts.py]: This should be a date and if the date is irrelevant, use a constant like `date.min`

File: fava/core/charts.py [hunk 2]
```diff
@@ -165,11 +165,16 @@ def net_worth(self, interval):
...
+            hour=0, minute=0, second=0, microsecond=0)
         for date in self.ledger.interval_ends(interval):
             while txn and txn.date < date:
                 for posting in filter(lambda p: p.account.startswith(types),
                                       txn.postings):
-                    inventory.add_position(posting)
+                    inventory.add_amount(
+                        posting.units,
+                        Cost(0, posting.cost.currency, today, None)
```
Comment (yagebu) [inline fava/core/charts.py]: The 0 should be a `Decimal`, so use the constant from `beancount.core.number`

File: fava/core/charts.py [hunk 3]
```diff
@@ -165,11 +165,16 @@ def net_worth(self, interval):
...
         inventory = Inventory()
 
+        today = datetime.datetime.today().replace(
+            hour=0, minute=0, second=0, microsecond=0)
         for date in self.ledger.interval_ends(interval):
             while txn and txn.date < date:
                 for posting in filter(lambda p: p.account.startswith(types),
                                       txn.postings):
-                    inventory.add_position(posting)
+                    inventory.add_amount(
```
Comment (yagebu) [inline fava/core/charts.py]: Maybe a short comment here why we do the calculations in this slightly odd way would be good.

File: fava/core/charts.py [hunk 4]
```diff
@@ -50,16 +50,25 @@ def _inventory_cost_or_value(inventory, date):
...
+    """Combine two inventory dicts adding numbers together -> amount dict."""
+    keys = set(inv1.keys()) | set(inv2.keys())
+    return {currency: inv1.get(currency, 0) + inv2.get(currency, 0)
+            for currency in keys}
+
+
 def _serialize_real_account(real_account, date):
+    children = [_serialize_real_account(account, date)
+                for _, account in sorted(real_account.items())]
+    balance_children = _inventory_cost_or_value(real_account.balance, date)
```
Comment (yagebu) [inline fava/core/charts.py]: If we use a `collections.Counter` here, this code should get much simpler

**Comment by yagebu** [comment]
It would be good to add tests for the functions that were changed.

**Comment by xentac** [comment]
I'll try to write tests, but here's the other requested changes.

File: fava/core/charts.py (hunk 1 above)
Comment (xentac) [reply fava/core/charts.py]: Whoops, you're right. Done.

File: fava/core/charts.py (hunk 3 above)
Comment (xentac) [reply fava/core/charts.py]: Done.

File: fava/core/charts.py (hunk 2 above)
Comment (xentac) [reply fava/core/charts.py]: Done.

File: fava/core/charts.py (hunk 4 above)
Comment (xentac) [reply fava/core/charts.py]: Neat, I didn't even know that existed.

**Comment by xentac** [comment]
Tests added.

**Comment by yagebu** [comment]
Merged with some minor changes, thanks 👍 

For me, this PR doesn't change the performance of the balance sheets very much, the improvement is at most 5-10%.

**Comment by xentac** [comment]
For reference, the reason mine is so large is because I started using Betterment in 2015 and contributed to it every other week. My Holdings report has 2005 rows in it...

**Pull request closed without merging**